src/cpp/modules/manualEigen/lstm_eigen_d.cpp

```diff
@@ -0,0 +1,366 @@
+#include "lstm_eigen_d_structures.h"
+#include "../../shared/lstm_eigen.h"
+
+// UTILS
+
+// Sigmoid diff on vector
+template<typename T>
+ArrayX<T> sigmoid_d(const ArrayX<T>& x, ArrayX<T>& d) {
+    ArrayX<T> s = sigmoid(x);
+    d.resize(s.rows());
+    d = s * (1 - s);
+    return s;
+}
+
+// tanh diff on vector
+template<typename T>
+ArrayX<T> tanh_d(const ArrayX<T>& x, ArrayX<T>& d) {
+    ArrayX<T> t = tanh(x);
+    d.resize(t.rows());
+    d = 1 - t * t;
+    return t;
+}
+
+// value (returned) and gradient (written to J) of log(sum(exp(x), 2))
+template<typename T>
+T logsumexp_d(const ArrayX<T>& vect, ArrayX<T>& J) {
+    J = exp(vect);
+    T sum = J.sum();
+    sum += 2.;
+
+    J /= sum;
+    return log(sum);
+}
+
+template<typename T>
+void swap(StateJacobianPredict<T>& j1, StateJacobianPredict<T>& j2)
+{
+    j1.layer.swap(j2.layer);
+}
+
+// OBJECTIVE
+
+// Manual Jacobian of lstm_model
+// Outputs jacobian containing the derivatives of the new state
+// with relation to params, state, and input
+template<typename T>
+void lstm_model_d(int hsize,
+    const LayerParams<T>& params,
+    LayerState<T>& state,
+    const ArrayX<T>& input,
+    ModelJacobian<T>& jacobian)
+{
+    ArrayX<T> forget_in(input * params.weight.forget + params.bias.forget);
+    ArrayX<T> forget_sd;
+    ArrayX<T> forget(sigmoid_d(forget_in, forget_sd));
+    ArrayX<T> forget_dw(forget_sd * input);
+    ArrayX<T> forget_db(forget_sd);
+    ArrayX<T> forget_di(forget_sd * params.weight.forget);
+
+    ArrayX<T> ingate_in(state.hidden * params.weight.ingate + params.bias.ingate);
+    ArrayX<T> ingate_sd;
+    ArrayX<T> ingate(sigmoid_d(ingate_in, ingate_sd));
+    ArrayX<T> ingate_dw(ingate_sd * state.hidden);
+    ArrayX<T> ingate_db(ingate_sd);
+    ArrayX<T> ingate_dh(ingate_sd * params.weight.ingate);
+
+    ArrayX<T> outgate_in(input * params.weight.outgate + params.bias.outgate);
+    ArrayX<T> outgate_sd;
+    ArrayX<T> outgate(sigmoid_d(outgate_in, outgate_sd));
+    ArrayX<T> outgate_dw(outgate_sd * input);
+    ArrayX<T> outgate_db(outgate_sd);
+    ArrayX<T> outgate_di(outgate_sd * params.weight.outgate);
+
+    ArrayX<T> change_in(state.hidden * params.weight.change + params.bias.change);
+    ArrayX<T> change_td;
+    ArrayX<T> change(tanh_d(change_in, change_td));
+    ArrayX<T> change_dw(change_td * state.hidden);
+    ArrayX<T> change_db(change_td);
+    ArrayX<T> change_dh(change_td * params.weight.change);
+
+    // Cell derivatives
+    ArrayX<T> orig_cell = state.cell;
+    state.cell = orig_cell * forget + ingate * change;
+    // wrt weight
+    jacobian.cell.d_weight_forget = orig_cell * forget_dw;
+    jacobian.cell.d_weight_ingate = change * ingate_dw;
+    jacobian.cell.d_weight_outgate = 0.;
+    jacobian.cell.d_weight_change = ingate * change_dw;
+    // wrt bias
+    jacobian.cell.d_bias_forget = orig_cell * forget_db;
+    jacobian.cell.d_bias_ingate = change * ingate_db;
+    jacobian.cell.d_bias_outgate.setZero();
```

Comment:
Why do we use use `= 0.` for `d_weight_outgate` and `.setZero()` for `d_bias_outgate`? Is there any difference?

Reply:
No, its not. I've just overlooked that.
Fixed now.